### PR TITLE
#170947548 Login with Facebook and Google

### PR DESCRIPTION
**Pull request by Ntare22**
#### What does this PR do?
- Enables users to sign with their google or facebook accounts

#### Description of Task to be completed?
- Install passport, passport-facebook, and passport-google-oauth
- Setup apps on google developer console and facebook facebook developer
- Write passport configuration file with facebook and google
- create a getProfile callback function.
- Setup routes for google and facebook redirecting
- Initialise passport in index.js

#### How should this be manually tested?
- Clone repository 
- Run `npm install`
- Open localhost link provided in terminal. ex `http://localhost:8000/api`
- Then open this link `http://localhost:8000/api/v1/auth/google` or `facebook`
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
[#170947548](https://www.pivotaltracker.com/story/show/170947548)
#### Screenshots (if appropriate)
image
image
image

### CODE REVIEW

File: src/routes/authRoutes.js [hunk 1]
```diff
@@ -1,13 +1,32 @@
...
+    'https://www.googleapis.com/auth/userinfo.profile',
+    'https://www.googleapis.com/auth/userinfo.email'
+  ]
+}));
+
+authRouter.get('/google/redirect', passport.authenticate('google'), AuthController.oAuthLogin);
+
+
+authRouter.get('/facebook', passport.authenticate('facebook', {
+  scope: ['emails'],
```
Comment (Mnickii) [inline src/routes/authRoutes.js]: why do we have this scope?

**Review by Mnickii — CHANGES_REQUESTED** [review]

File: src/seeders/20200212121323-User.js [hunk 2]
```diff
@@ -4,8 +4,17 @@ module.exports = {
   up: (queryInterface) => queryInterface.bulkInsert(
     'Users',
     [
+      // {
+      //   id: 'd0651d02-46e6-4b70-9c3b-8d378c178241',
+      //   firstName: 'bienaime',
+      //   lastName: 'jeanb',
+      //   email: 'aime@andela.com',
+      //   password: bcrypt.hashSync('Aime12&*', Number(process.env.passwordHashSalt)),
```
Comment (Mnickii) [inline src/seeders/20200212121323-User.js]: remove commented code

**Review by Mnickii — CHANGES_REQUESTED** [review]

**Review by izzett222 — DISMISSED** [review]

File: src/controllers/authController.js [hunk 3]
```diff
@@ -89,6 +89,53 @@ export default class AuthController {
...
+      const { id } = profile;
+      const firstName = profile.name.givenName;
+      const lastName = profile.name.familyName;
+      const email = Array.isArray(profile.emails) ? profile.emails[0].value : '';
+
+      const oAuthId = req.user.id;
+      const isVerified = true;
+      const password = 'null';
+
+      // const email = Array.isArray(emails) && emails[0] && emails[0].value;
```
Comment (Mnickii) [inline src/controllers/authController.js]: commented code

File: src/controllers/authController.js [hunk 4]
```diff
@@ -89,6 +89,52 @@ export default class AuthController {
...
+     * @memberof authController
+     */
+  static async oAuthLogin(req, res) {
+    try {
+      const profile = { ...req.user };
+      const method = profile.provider;
+      const { id } = profile;
+      const firstName = profile.name.givenName;
+      const lastName = profile.name.familyName;
+      const email = Array.isArray(profile.emails) ? profile.emails[0].value : '';
```
Comment (Mnickii) [inline src/controllers/authController.js]: Do we need to check specifically if it's an array or if the profile.emails are present? If the latter, ```profile.emails ? profile.emails[0].value : ''``` should do

File: src/seeders/20200212121323-User.js [hunk 5]
```diff
@@ -6,7 +6,7 @@ module.exports = {
     'Users',
     [
       {
-        id: uuid(),
+        id: "d0551d02-46e6-4b79-9c3b-8d378c178243",
```
Comment (Mnickii) [inline src/seeders/20200212121323-User.js]: why this change?

File: src/tests/oAuth.test.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      });
+  });
+});
+
+
+describe('Facebook oAuthentication tests', () => {
+  it('should return user is successfully created', (done) => {
+    chai
+      .request(app)
+      .post('/api/v1/auth/facebook')
```
Comment (Mnickii) [inline src/tests/oAuth.test.js]: it's not efficient to send unwanted requests to external APIs. Remember we have rate limiting and the way you've done it, we hit the external APIs everytime someone runs the tests

**Review by Mnickii — CHANGES_REQUESTED** [review]

File: src/tests/oAuth.test.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        name: { familyName: 'Ntare', givenName: 'Jim' },
+        emails: [{ value: 'jim.ntare@gmail.com', verified: true }],
+        provider: 'google',
+      }
+    };
+    const res = { status: () => ({ json: (data) => ({ body: { ...data } }) }) };
+    const results = await authController.oAuthLogin(req, res);
+    expect(results.body.status).to.equal(200);
+    expect(results.body.message).to.equal('User is successfully created');
+    // console.log('>>>', results);
```
Comment (Mnickii) [inline src/tests/oAuth.test.js]: commented code.

File: src/tests/index.test.js [hunk 8]
```diff
@@ -20,16 +19,3 @@ describe('Testing index page ', () => {
       });
   });
 });
-
-describe('Testing index page ', () => {
-  it('should return a message from index page in French.', (done) => {
```
Comment (Mnickii) [inline src/tests/index.test.js:25]: why remove this test?

File: src/seeders/20200212121323-User.js (hunk 5 above)
Comment (Ntare22) [reply src/seeders/20200212121323-User.js]: It Baraka who put the uuid().... which might mean I didn't rebase well.

File: src/controllers/authController.js (hunk 4 above)
Comment (Ntare22) [reply src/controllers/authController.js]: I changed that.

File: src/seeders/20200212121323-User.js (hunk 5 above)
Comment (Mnickii) [reply src/seeders/20200212121323-User.js]: Oh, that should be it. It seemed as though you'd undone Baraka's changes

File: src/tests/4-verifyAccount.test.js [hunk 9]
```diff
@@ -43,7 +43,7 @@ describe('Verification tests', () => {
     chai.request(index)
       .get(`/api/v1/auth/verification?token=${wrongToken}&email=${email}`)
       .end((err, res) => {
-        expect(res.body).to.be.an('object');
+        expect(res.body).to.be.a('object');
```
Comment (Mnickii) [inline src/tests/4-verifyAccount.test.js]: this change isn't necessary

File: src/tests/index.test.js (hunk 8 above)
Comment (Ntare22) [reply src/tests/index.test.js:25]: It is not removed.... we rename the testing documents...!

File: src/tests/5-oAuth.test.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    expect(results.body.status).to.equal(200);
+    expect(results.body.message).to.equal('User is successfully logged in');
+  });
+});
+
+describe('FACEBOOK oAuthentication tests', (done) => {
+  it('should redirect to facebook', async () => {
+    chai
+      .request(index)
+      .get('/api/v1/auth/facebook')
```
Comment (Mnickii) [inline src/tests/5-oAuth.test.js]: let's not hit external APIs at all. mock the requests

File: src/tests/1-index.test.js [hunk 11]
```diff
@@ -20,16 +19,3 @@ describe('Testing index page ', () => {
...
   });
 });
-
-describe('Testing index page ', () => {
-  it('should return a message from index page in French.', (done) => {
-    chai.request(app)
-      .get('/api?lang=fr')
-      .end((err, res) => {
-        expect(res.body).to.be.an('object');
-        expect(res.status).to.equal(200);
```
Comment (Mnickii) [inline src/tests/1-index.test.js]: this test is necessary

File: src/tests/5-oAuth.test.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        name: { familyName: 'Ntare', givenName: 'Jim' },
+        emails: [{ value: 'jim.ntare@gmail.com', verified: true }],
+        provider: 'google',
+      }
+    };
+    const res = { status: () => ({ json: (data) => ({ body: { ...data } }) }) };
+    const results = await authController.oAuthLogin(req, res);
+    expect(results.body.status).to.equal(200);
+    expect(results.body.message).to.equal('User is successfully created');
+    // console.log('>>>', results);
```
Comment (Mnickii) [inline src/tests/5-oAuth.test.js]: remove commented code

File: src/services/factory/googleUser.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import Chance from 'chance';
+import { Factory } from 'rosie';
+
+const chance = new Chance();
```
Comment (Mnickii) [inline src/services/factory/googleUser.js]: we don't need these anymore

File: src/services/factory/facebookUser.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import Chance from 'chance';
+import { Factory } from 'rosie';
+
+const chance = new Chance();
+
+export default Factory.define('user')
+  .attr('id', `${chance.integer({ min: 100, max: 100000 })}`)
```
Comment (Mnickii) [inline src/services/factory/facebookUser.js]: this too

**Review by Baraka-Mugisha — DISMISSED** [review]

**Review by izzett222 — DISMISSED** [review]

File: package.json [hunk 15]
```diff
@@ -33,17 +34,23 @@
...
     "bcrypt": "^3.0.8",
     "body-parser": "^1.19.0",
+    "cookie-session": "^1.4.0",
     "dotenv": "^6.2.0",
     "express": "^4.17.1",
     "express-validator": "^6.4.0",
     "i18n": "^0.8.5",
     "jsonwebtoken": "^8.5.1",
     "localStorage": "^1.0.4",
+    "node-mocks-http": "^1.8.1",
```
Comment (Mnickii) [inline package.json]: is this used somewhere? let's not include unused packages,

**Review by Mnickii — CHANGES_REQUESTED** [review]

File: package.json (hunk 15 above)
Comment (Ntare22) [reply package.json]: I've removed it.

**Review by Baraka-Mugisha — DISMISSED** [review]

**Review by Baraka-Mugisha — DISMISSED** [review]

File: src/tests/3-user.test.js [hunk 16]
```diff
@@ -21,7 +21,7 @@ describe('USER ROLES TESTS', () => {
       })
       .end((err, res) => {
         token = res.body.data;
-        expect(res.status).to.equal(200);
+        expect(res.status).equal(200);
```
Comment (Mnickii) [inline src/tests/3-user.test.js]: return this

File: src/tests/3-user.test.js [hunk 17]
```diff
@@ -36,7 +36,7 @@ describe('USER ROLES TESTS', () => {
       .end((err, res) => {
         expect(res.status).to.equal(200);
         unauthToken = res.body.data;
-
+        expect(res.status).equal(200);
```
Comment (Mnickii) [inline src/tests/3-user.test.js]: remove this

**Review by izzett222 — APPROVED** [review]

**Review by Mnickii — APPROVED** [review]
LGTM

**Review by Baraka-Mugisha — APPROVED** [review]
